common/rangeiter/dynamic_config_linear_iterator.go

```diff
@@ -0,0 +1,123 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package rangeiter
+
+import "sync"
+
+// DynamicConfigLinearIterator is a LinearIterator that can be reconfigured dynamically.
+// It is safe for concurrent use. Every operation checks the values of min, max, and stepCount in dynamic config
+// and refreshes the iterator if they have changed.
+type DynamicConfigLinearIterator[T Integer] struct {
+	mx sync.Mutex
+
+	iter    *LinearIterator[T]
+	curSpec *linearIteratorSpec[T]
+
+	minFn       func() T
+	maxFn       func() T
+	stepCountFn func() int
+}
+
+// NewDynamicConfigLinearIterator creates a new DynamicConfigLinearIterator.
+func NewDynamicConfigLinearIterator[T Integer](min, max func() T, stepCount func() int) *DynamicConfigLinearIterator[T] {
+	iter := &DynamicConfigLinearIterator[T]{
+		minFn:       min,
+		maxFn:       max,
+		stepCountFn: stepCount,
+	}
+	iter.curSpec = iter.newLinearIteratorSpec()
+	iter.iter = iter.curSpec.newIterator()
+	return iter
+}
+
+// refreshIterator checks if the spec has changed and refreshes the iterator if it has.
+func (d *DynamicConfigLinearIterator[T]) refreshIterator() {
+	newSpec := d.newLinearIteratorSpec()
+	if newSpec.equal(d.curSpec) {
+		return
+	}
+
+	d.curSpec = newSpec
```

Comment:
let's add an info log here including the spec content so we have visibility when it reloads and what the current spec is